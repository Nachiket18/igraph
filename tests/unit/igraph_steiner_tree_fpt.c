/*
   IGraph library.
   Copyright (C) 2022  The igraph development team <igraph@igraph.org>

   This program is free software; you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation; either version 2 of the License, or
   (at your option) any later version.

   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with this program.  If not, see <https://www.gnu.org/licenses/>.
*/

#include <igraph.h>
#include "test_utilities.h"

void check_graph(const igraph_t *graph, const igraph_vector_int_t *terminals, const igraph_vector_t *weights) {
    igraph_real_t value;
    igraph_vector_int_t tree_edges;

    igraph_vector_int_init(&tree_edges, 0);

    igraph_error_t x = igraph_steiner_dreyfus_wagner(graph, terminals, weights, &value, &tree_edges);
    IGRAPH_ASSERT(x == IGRAPH_SUCCESS);

    print_vector_int(&tree_edges);
    printf("Total Steiner tree weight: %g\n", value);

    /* Check total tree weight. */
    igraph_real_t value2 = 0.0;
    igraph_integer_t tree_size = igraph_vector_int_size(&tree_edges);
    for (igraph_integer_t i = 0; i < tree_size; i++) {
        value2 += VECTOR(*weights)[ VECTOR(tree_edges)[i] ];
    }
    IGRAPH_ASSERT(value == value2);

    /* Check that the result is indeed a tree. */
    if (igraph_vector_int_size(terminals) > 0) {
        igraph_t tree;
        igraph_subgraph_from_edges(graph, &tree, igraph_ess_vector(&tree_edges), /* delete_vertices= */ true);

        igraph_bool_t is_tree;
        igraph_is_tree(&tree, &is_tree, NULL, IGRAPH_ALL);
        IGRAPH_ASSERT(is_tree);

        igraph_destroy(&tree);
    }

    igraph_vector_int_destroy(&tree_edges);
}

int main(void) {
    igraph_t g_null, g_k7, g_k6_k1, g_k7_n, g_k7_n1, g_k7_real, g_k7_non_simple,g_k1,g_k7_self_loop;
    igraph_vector_int_t terminals_null, terminals_k7, terminals_k6_k1, terminals_k7_real;
    igraph_vector_t weights_null, weights_k7, weights_k6_k1, weights_k7_n, weights_k7_n1, weights_k7_real, weights_k7_non_simple,weights_k7_loop;

    /* Null graph */
    igraph_empty(&g_null, 0, 0);
    igraph_vector_init(&weights_null, 0);
    igraph_vector_int_init(&terminals_null, 0);

    /* K_7 complete graph with a specific edge ordering. */
    igraph_small(&g_k7, 7, IGRAPH_UNDIRECTED,
                 0, 1, 0, 2, 0, 3, 0, 4, 0, 5, 0, 6,
                 1, 2, 1, 3, 1, 4, 1, 5, 1, 6,
                 2, 3, 2, 4, 2, 5, 2, 6,
                 3, 4, 3, 5, 3, 6,
                 4, 5, 4, 6,
                 5, 6,
                 -1);
    // 1 , 1
    igraph_vector_init_int(&weights_k7, 21,
                           2, 2, 2, 1, 1, 2,
                           2, 2, 2, 1, 2,
                           2, 2, 2, 1,
                           1, 2, 1,
                           2, 1,
                           1);

    igraph_vector_int_init_int(&terminals_k7, 4,
                               0, 1, 2, 3);


    /* K_7 non-complete graph with edges 0-4,0-5 removed */
    igraph_small(&g_k7_n, 7, IGRAPH_UNDIRECTED,
                 0, 1, 0, 2, 0, 3, 0, 6,
                 1, 2, 1, 3, 1, 4, 1, 5, 1, 6,
                 2, 3, 2, 4, 2, 5, 2, 6,
                 3, 4, 3, 5, 3, 6,
                 4, 5, 4, 6,
                 5, 6,
                 -1);

    igraph_vector_init_int(&weights_k7_n, 19,
                           2, 2, 2, 2,
                           2, 2, 2, 1, 2,
                           2, 2, 2, 1,
                           1, 2, 1,
                           2, 1,
                           1);


    /* K_7 non-complete graph with edges 0-4 removed */
    igraph_small(&g_k7_n1, 7, IGRAPH_UNDIRECTED,
                 0, 1, 0, 2, 0, 3, 0, 4, 0, 6,
                 1, 2, 1, 3, 1, 4, 1, 5, 1, 6,
                 2, 3, 2, 4, 2, 5, 2, 6,
                 3, 4, 3, 5, 3, 6,
                 4, 5, 4, 6,
                 5, 6,
                 -1);

    igraph_vector_init_int(&weights_k7_n1, 20,
                           2, 2, 2, 1, 2,
                           2, 2, 2, 1, 2,
                           2, 2, 2, 1,
                           1, 2, 1,
                           2, 1,
                           1);


    /* Disconnected graph: K_6 with a specific edge order and an additional isolated vertex. */
    igraph_small(&g_k6_k1, 7, IGRAPH_UNDIRECTED,
                 0, 1, 0, 2, 0, 3, 0, 4, 0, 5,
                 1, 2, 1, 3, 1, 4, 1, 5,
                 2, 3, 2, 4, 2, 5,
                 3, 4, 3, 5,
                 4, 5,
                 -1);

    igraph_vector_init_int(&weights_k6_k1, 15,
                           2, 2, 2, 1, 1,
                           2, 2, 2, 1,
                           2, 2, 2,
                           1, 2,
                           2);

    igraph_vector_int_init_int(&terminals_k6_k1, 4,
                               0, 1, 2, 6);

    igraph_small(&g_k7_real, 7, IGRAPH_UNDIRECTED,
                 0, 1, 0, 2, 0, 3, 0, 4, 0, 5, 0, 6,
                 1, 2, 1, 3, 1, 4, 1, 5, 1, 6,
                 2, 3, 2, 4, 2, 5, 2, 6,
                 3, 4, 3, 5, 3, 6,
                 4, 5, 4, 6,
                 5, 6,
                 -1);
    // 1 , 1
    igraph_vector_init_real(&weights_k7_real, 21,
                            2.5, 2.5, 2.5, 0.5, 0.5, 2.5,
                            2.5, 2.5, 2.5, 0.5, 2.5,
                            2.5, 2.5, 2.5, 0.5,
                            0.5, 2.5, 0.5,
                            2.5, 0.5,
                            0.5);

    igraph_vector_int_init_int(&terminals_k7_real, 4,
                               0, 1, 2, 3);


    /* K_7 complete graph with a specific edge ordering. */
    igraph_small(&g_k7_non_simple, 7, IGRAPH_UNDIRECTED,
                 0, 1, 0, 2, 0, 3, 0, 4, 0, 5, 0, 5, 0, 6,
                 1, 2, 1, 3, 1, 4, 1, 5, 1, 6,
                 2, 3, 2, 4, 2, 5, 2, 6,
                 3, 4, 3, 5, 3, 6,
                 4, 5, 4, 6,
                 5, 6,
                 -1);
    // 1 , 1
    igraph_vector_init_int(&weights_k7_non_simple, 22,
                           2, 2, 2, 1, 1, 2, 2,
                           2, 2, 2, 1, 2,
                           2, 2, 2, 1,
                           1, 2, 1,
                           2, 1,
                           1);

    // A graph with 1 node.
    igraph_small(&g_k1, 1, IGRAPH_UNDIRECTED,-1);
    

    /* 
        K_7 graph with a specific edge ordering and a self loop 
        edge 0-1 is removed so it's not a complate graph
    */
    igraph_small(&g_k7_self_loop, 7, IGRAPH_UNDIRECTED,
                 0, 0, 0, 2, 0, 3, 0, 4, 0, 5, 0, 6,
                 1, 2, 1, 3, 1, 4, 1, 5, 1, 6,
                 2, 3, 2, 4, 2, 5, 2, 6,
                 3, 4, 3, 5, 3, 6,
                 4, 5, 4, 6,
                 5, 6,
                 -1);
    // 1 , 1
    igraph_vector_init_int(&weights_k7_loop, 21,
                           2, 2, 2, 1, 1, 2,
                           2, 2, 2, 1, 2,
                           2, 2, 2, 1,
                           1, 2, 1,
                           2, 1,
                           1);

    igraph_vector_int_init_int(&terminals_k7, 4,
                               0, 1, 2, 3);



    printf("Null graph:\n");
    check_graph(&g_null, &terminals_null, &weights_null);

    printf("\nK_7 graph:\n");
    check_graph(&g_k7, &terminals_k7, &weights_k7);

    printf("\nK_7 Non Complete graph-1:\n");
    check_graph(&g_k7_n, &terminals_k7, &weights_k7_n);

    printf("\nK_7 Non Complete graph-2:\n");
    check_graph(&g_k7_n1, &terminals_k7, &weights_k7_n1);

    igraph_real_t value;
    igraph_vector_int_t tree_edges;
    igraph_vector_int_init(&tree_edges, 0);

    CHECK_ERROR(igraph_steiner_dreyfus_wagner(&g_k6_k1, &terminals_k6_k1, &weights_k6_k1, &value, &tree_edges), IGRAPH_EINVAL);

    igraph_vector_int_destroy(&tree_edges);


    igraph_real_t value_unit;
    igraph_vector_int_t tree_edges_unit;
    igraph_vector_int_init(&tree_edges_unit, 0);

    printf("\nK_7_Unit graph:\n");
    igraph_steiner_dreyfus_wagner(&g_k7, &terminals_k7, NULL, &value_unit, &tree_edges_unit);
    print_vector_int(&tree_edges_unit);
    printf("Total Steiner tree weight: %g\n", value_unit);

    igraph_vector_int_destroy(&tree_edges_unit);

    printf("\nK_7_Real graph:\n");
    check_graph(&g_k7_real, &terminals_k7_real, &weights_k7_real);

    printf("\nK_7 graph with parallel edges:\n");
    check_graph(&g_k7_non_simple, &terminals_k7, &weights_k7_non_simple);

    printf("\nA graph with single node:\n");
    check_graph(&g_k1,&terminals_null,&weights_null);

    printf("\nK_7 graph with self loop:\n");
    check_graph(&g_k7_self_loop,&terminals_k7,&weights_k7_loop);


    igraph_destroy(&g_null);
    igraph_destroy(&g_k7);
    igraph_destroy(&g_k6_k1);
    igraph_destroy(&g_k7_real);
    igraph_destroy(&g_k7_n1);
    igraph_destroy(&g_k7_n);
    igraph_destroy(&g_k7_non_simple);
    

    igraph_vector_destroy(&weights_null);
    igraph_vector_destroy(&weights_k7);
    igraph_vector_destroy(&weights_k6_k1);
    igraph_vector_destroy(&weights_k7_n);
    igraph_vector_destroy(&weights_k7_n1);
    igraph_vector_destroy(&weights_k7_real);
    igraph_vector_destroy(&weights_k7_non_simple);

    igraph_vector_int_destroy(&terminals_k7);
    igraph_vector_int_destroy(&terminals_null);
    igraph_vector_int_destroy(&terminals_k6_k1);
    igraph_vector_int_destroy(&terminals_k7_real);

    VERIFY_FINALLY_STACK();

    return 0;
}
