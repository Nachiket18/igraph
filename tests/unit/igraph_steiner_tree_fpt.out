Null graph:
( )
Total Steiner tree weight: 0

K_7 graph:
( 4 9 14 17 20 )
Total Steiner tree weight: 5

K_7 Non Complete graph-1:
( 0 1 2 )
Total Steiner tree weight: 6

K_7 Non Complete graph-2:
( 0 1 2 )
Total Steiner tree weight: 6

K_7_Unit graph:
( 0 1 2 )
Total Steiner tree weight: 3

K_7_Real graph:
( 4 9 14 17 20 )
Total Steiner tree weight: 2.5

K_7 graph with parallel edges:
( 4 10 15 18 21 )
Total Steiner tree weight: 5

A graph with single node:
( )
Total Steiner tree weight: 0

K_7 graph with self loop:
( 4 9 14 17 20 )
Total Steiner tree weight: 5

A graph with n-1 terminals:
0 1 2 3 4 5 6 7
value: 8.000000

A graph with 1 terminal:

value: 0.000000

A graph with 2 terminals:
12
value: 1.000000

A graph with 3 terminals:
10 12
value: 2.000000

A graph:
Tree edges:
0 1 2 3 4 5 6 7
value: 8.000000

A graph with 8 terminals:
Tree edges:
1 2 3 4 5 6 7
value: 7.000000

A graph with 7 terminals and 2 removed to test if it forces it to go through:
Tree edges:
1 2 3 4 5 6 7
value: 7.000000

A graph with different structure that previous
Tree edges:
0 1 2 4 6 7
value: 6.000000

A simple square graph with few more edges outside the square
Tree edges:
0 2 3 5
value: 4.000000

A 3 vertices square graph with weights
Tree edges:
0 1 3 8 11
value: 500.000000
